### ISSUE TITLE
Followups to #66

### ISSUE THREAD

**Issue by adriangb** [adriangb/scikeras#79]
Possible followups from discussion in #66 

- [ ] Adding regex and/or globbing based parameter routing using a `__params_group` suffix.
- [x] Catching user errors when users compile their own models (https://github.com/adriangb/scikeras/pull/66#discussion_r492362306, done in #86)
- [ ] More user-friendly warnings when parameters are passed to an item that is not a class (https://github.com/adriangb/scikeras/pull/66#discussion_r492936838)
- [x] A bidirectional mapping between shorthand names for losses/metrics/optimizers and their functions/classes (https://github.com/adriangb/scikeras/pull/66#discussion_r492863842) Unidirectional mapping implemented in #88.
- [x] Consider making utilities public and refactor testing to remove tests for private functions.

**Comment by stsievert** [comment]
Can you add making "Make some utilities public, and remove tests of private functions"?

**Comment by adriangb** [comment]
I looked around, it looks like the only place we are testing "private" functions is within `test_utils.py`. It should be easy enough to replace them. Are there any specific utilities or private methods you think we should make public?

Regarding catching of user errors when compiling their models, would you be open to making this a warning instead of an error? I feel that it still achieves the goal of letting the user know they may have made a mistake while not going so far as to prohibiting users from doing things we don't approve of or haven't considered necessary.

**Comment by stsievert** [comment]
> Are there any specific utilities or private methods you think we should make public?

I think the bidirectional map(s) between losses/optimizers/metrics should be public. Maybe `LabelDimensionTransformer` if users can (easily) customize BaseWrapper to use a different class. I'm not seeing why users would use any other utilities.

> making this a warning instead of an error? I feel that it still achieves the goal of letting the user know they may have made a mistake while not going so far as to prohibiting users from doing things we don't approve of or haven't considered necessary.

Generally, I tend to raise errors with helpful messages (or at least I try). Every time I write code to raise an exception I think "how could the user clear resolve this issue?" and include those actions in the error message (e.g,  https://github.com/adriangb/scikeras/pull/66#issuecomment-696750092).

**Comment by adriangb** [comment]
> Maybe LabelDimensionTransformer if users can (easily) customize BaseWrapper to use a different class

This is an interesting one. As per #78 I think it may be beneficial to rethink how we handle the data transformations. Currently users are able to override `preprocess_{X,y}` and/or `postprocess_{X,y}`. There is also `_check_output_model_compatibility` that also transforms data but runs _after_ the `model_` is compiled because it uses information about the loss functions to reshape the data. This is a bit convoluted. Perhaps it's better to have a transformer-like class with methods like `process_{X,y}_pre_build`, `process_{X,y}_post_build` and `process_{X,y}_post_predict` and `get_meta_params` (or better names) instead of methods within the wrappers themselves. This would also absorb the work that `LabelDimensionTransformer ` currently does for KerasClassifier. That way users just have to do something like:

```python3
est = KerasClassifier(model=get_model, data_processor=DataProcessorClass())
```

**Comment by adriangb** [comment]
> Can you add making "Make some utilities public, and remove tests of private functions"?

Is there anything specific you had in mind?

**Comment by stsievert** [comment]
As far as I can tell, all the tests of private functions have been removed. :+1:

I think TFRandomState should be public. What if a user wants to initialize all the models the same but ignore any performance penalties with the CUDA keywords?

> Setting the environment variable [TF_DETERMINISTIC_OPS=1] forces the selection of deterministic cuDNN convolution and max-pooling algorithms. ...
>
> Selecting these deterministic options may reduce performance. 
>
> —https://docs.nvidia.com/deeplearning/frameworks/tensorflow-release-notes/rel_19.06.html#key-features-and-enhancements

I have some more ideas for utilities in https://github.com/adriangb/scikeras/issues/83#issuecomment-701806316.

**Comment by adriangb** [comment]
> I think TFRandomState should be public

TFRandomState is a temporary solution. From discussion with the TF team, in the future they will adopt [tf.random.Generator](https://www.tensorflow.org/api_docs/python/tf/random/Generator), which should make it possible to handle things in a much less hacky way.

**Comment by adriangb** [comment]
> A bidirectional mapping between shorthand names mapping between shorthand names for losses/metrics/optimizers and their functions/classes

We now have a shorthand -> class/func implementation in #88. We don't have any direct need to the other direction, and I think there would be a lot of issues around handling edge cases, synonyms (ex: acc and accuracy and categorical_accuracy). I'd say for now we hold off on this until we either need it or users request it.

**Comment by stsievert** [comment]
> TFRandomState is a temporary solution

Perfect. It's really hacky, but it does provide utility.

**Comment by adriangb** [comment]
The main issue left here is the globbing/regex based parameter routing. I opened #118 to track that on a clean slate. I'm going to close this issue in favor of that one.
